swim/heal_via_discover_provider.go

```diff
@@ -0,0 +1,166 @@
+// the discoverProvider for nodes that are faulty or not available in its
+// membership. It attempts the heal probabalisticly so that the discovery provider
+// doesn't get overloaded with request -- with default settings 6
+// times per minutes on avarage for the entire cluster.
+type discoverProviderHealer struct {
+	node *Node
+
+	discoverProvider *discovery.DiscoverProvider
+	baseProbabillity float64
+	period           time.Duration
+
+	previousHostListSize int
+	quit                 chan struct{}
+
+	logger log.Logger
+}
+
+func newDiscoverProviderHealer(n *Node, baseProbability float64, period time.Duration) *discoverProviderHealer {
+	return &discoverProviderHealer{
+		node:             n,
+		baseProbabillity: baseProbability,
+		period:           period,
+		quit:             make(chan struct{}),
+		logger:           logging.Logger("healer").WithField("local", n.Address()),
+	}
+}
+
+// Start the partition healing loop
+func (h *discoverProviderHealer) Start() {
+	go func() {
+		for {
+			// attempt heal with the pro
+			if rand.Float64() < h.Probability() {
+				h.Heal()
+			}
+
+			// loop or quit
+			select {
+			case <-h.node.clock.After(h.period):
+			case <-h.quit:
+				return
+			}
+		}
+	}()
+}
+
+// Stop the partition healing loop.
+func (h *discoverProviderHealer) Stop() {
+	select {
+	case <-h.quit:
+	default:
+		close(h.quit)
+	}
+}
+
+// Probability returns the probability when a heal should be attempted
+// we want to throttle the heal attempts to alleviate pressure on the
+// discover provider.
+func (h *discoverProviderHealer) Probability() float64 {
+	// avoid division by zero.
+	if h.previousHostListSize < h.node.CountReachableMembers() {
+		h.previousHostListSize = h.node.CountReachableMembers()
+	}
+	if h.previousHostListSize < 1 {
+		h.previousHostListSize = 1
+	}
+	return h.baseProbabillity / float64(h.previousHostListSize)
+}
+
+// Heal iterates over the hostList that the discoverProvider provides. If the
+// node encounters a host that is faulty or not in the membership, we pick that
+// node as a target to perform a partition heal with.
+//
+// If heal was attempted, returns identities of the target nodes.
+func (h *discoverProviderHealer) Heal() []string {
+	h.node.emit(DiscoHealEvent{})
+	// get list from discovery provider
+	if h.node.discoverProvider == nil {
+		return []string{}
+	}
+	hostList, err := h.node.discoverProvider.Hosts()
+	if err != nil {
+		h.logger.Warn("unable to receive host list from discover provider")
+		return []string{}
+	}
+
+	h.previousHostListSize = len(hostList)
+
+	var targets []string
+	for _, address := range hostList {
+		m, ok := h.node.memberlist.Member(address)
+		if !ok || m.Status == Faulty {
+			targets = append(targets, address)
+		}
+	}
+
+	// filter hosts that we already know about and attempt to heal nodes that
+	// are complementary to the membership of this node.
+	var ret []string
+
+	for len(targets) != 0 {
```

Comment:
I am still not convinced this is the better way to go.

In case of single-node clusters that haven't seen the rest of the cluster before (the only realistic way that many single-node clusters can happen IMO), _they_ will start the healing processes separately, thus quickly joining to the "big" cluster. Unless there are more arguments, I think it's a bit premature to develop for that case.

@benfleis what's your take on it?